paperqa/llms.py

```diff
@@ -65,9 +65,26 @@ class LiteLLMEmbeddingModel(EmbeddingModel):
     name: str = Field(default="text-embedding-3-small")
     embedding_kwargs: dict = Field(default_factory=dict)
 
+    def _truncate_if_large(self, texts: list[str]) -> list[str]:
+        """Truncate texts if they are too large specifically for an openai model."""
+        if self.name not in {
+            "text-embedding-3-small",
+            "text-embedding-3-large",
+            "text-embedding-ada-002",
+        }:
+            return texts
+        max_tokens = 8192
+        maybe_too_large = max_tokens * 3
+        if any(len(t) > maybe_too_large for t in texts):
+            enct = tiktoken.get_encoding("cl100k_base")
+            enc_batch = enct.encode_ordinary_batch(texts)
+            return [enct.decode(t[:max_tokens]) for t in enc_batch]
+        return texts
+
     async def embed_documents(
         self, texts: list[str], batch_size: int = 16
     ) -> list[list[float]]:
+        texts = self._truncate_if_large(texts)
```

Comment:
I doubt this will practically affect retrieval much (since 8k is already so large), but I wonder if we should support splitting + embedding then taking the averaging back into a single embedding for a large document.

Reply:
Oh I understand what you're saying - I'm not sure that works though. Would have to experiment